includes/wikia/nirvana/WikiaView.class.php

```diff
@@ -114,36 +109,106 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
 			}
 
-			// First we look for BaseName_MethodName
-			$dirName = dirname( $wgAutoloadClasses[$controllerClass] );
-			$basePath = "{$dirName}/templates/{$controllerBaseName}_{$methodName}";
-			$templatePath = "{$basePath}.$extension";
-			$templateExists = file_exists( $templatePath );
+			$this->setTemplatePath( $templatePath );
+		}
+		wfProfileOut(__METHOD__);
+	}
 
-			// Fall back to ControllerClass_MethodName
-			if( !$templateExists ) {
-				$templatePath = "{$dirName}/templates/{$controllerClass}_{$methodName}.$extension";
-				$templateExists = file_exists( $templatePath );
-			}
+	/**
+	 * For legacy reasons we sometimes get a $controllerClass name that doesn't end with
+	 * 'Controller' or 'Service'.  Normalize to this form.
+	 *
+	 * @param $controllerClass
+	 *
+	 * @return string
+	 *
+	 * @throws WikiaException
+	 */
+	public static function normalizeControllerClass( $controllerClass ) {
+		$app = F::app();
+		$controllerBaseName = $app->getBaseName( $controllerClass );
+		if ( $app->isService( $controllerClass ) ) {
+			$controllerClass = $app->getServiceClassName( $controllerBaseName );
+		} else {
+			$controllerClass = $app->getControllerClassName( $controllerBaseName );
+		}
 
-			if( !$templateExists ) {
-				throw new WikiaException( "Template file not found: {$templatePath}" );
+		if ( empty( $app->wg->AutoloadClasses[$controllerClass] ) ) {
+			throw new WikiaException( "Invalid controller or service name: {$controllerClass}" );
+		}
+
+		return $controllerClass;
+	}
+
+	/**
+	 * Generates a list of possible template names ordered by preference
+	 *
+	 * @param $controllerClass
+	 * @param $methodName
+	 *
+	 * @return array
+	 */
+	public function getTemplateOptions( $controllerClass, $methodName ) {
+		$templates = [];
+
+		// Make sure this method exists, otherwise the call to getMethod crashes PHP
+		// so badly it can't even log that a problem occurred.
+		if ( method_exists( $controllerClass, $methodName ) ) {
+			// See if there is a @template annotation for the method we're generating a view for
+			$reflection = new ReflectionClass( $controllerClass );
+			$method = $reflection->getMethod( $methodName );
+
+			$comment = $method->getDocComment();
+			if ( preg_match( '/@template ([^ ]+)/', $comment, $matches ) ) {
+				$templates[] = trim( $matches[ 1 ] );
 			}
+		}
 
-			$this->setTemplatePath( $templatePath );
+		// Add variations on the controller name
+		$controllerBaseName = F::app()->getBaseName( $controllerClass );
+		$templates[] = "{$controllerBaseName}_{$methodName}";
+		$templates[] = "{$controllerClass}_{$methodName}";
+
+		return $templates;
+	}
+
+	/**
+	 * See if the controller defines a custom template directory, otherwise use the default directory
+	 *
+	 * @param $controllerClass
+	 *
+	 * @return string
+	 */
+	public function getTemplateDir( $controllerClass ) {
+		if ( method_exists( $controllerClass, 'getTemplateDir' ) ) {
```

Comment:
I'd rather go with a dummy function returning null defined in base class so you can override it in a subclass if you need.